pex/scie/science.py

```diff
@@ -0,0 +1,320 @@
+                    science_binary_url
+                ) as read_fp:
+                    shutil.copyfileobj(read_fp, write_fp)
+
+                science_sha256_url = _science_binary_url(".sha256")
+                with fetcher.get_body_stream(science_sha256_url) as fp:
+                    expected_sha256, _, _ = fp.read().decode("utf-8").partition(" ")
+                actual_sha256 = CacheHelper.hash(target_science, hasher=Sha256)
+                if expected_sha256 != actual_sha256:
+                    raise ValueError(
+                        "The science binary downloaded from {science_binary_url} does not match "
+                        "the expected SHA-256 fingerprint recorded in {science_sha256_url}.\n"
+                        "Expected {expected_sha256} but found {actual_sha256}.".format(
+                            science_binary_url=science_binary_url,
+                            science_sha256_url=science_sha256_url,
+                            expected_sha256=expected_sha256,
+                            actual_sha256=actual_sha256,
+                        )
+                    )
+    return os.path.join(target_dir, "science")
+
+
+class ScienceError(Exception):
+    """Indicates an error executing science."""
+
+
+def build(
+    configuration,  # type: ScieConfiguration
+    pex_file,  # type: str
+    url_fetcher=None,  # type: Optional[URLFetcher]
+    env=ENV,  # type: Variables
+):
+    # type: (...) -> Iterator[ScieInfo]
+
+    science = _ensure_science(url_fetcher=url_fetcher, env=env)
+    name = re.sub(r"\.pex$", "", os.path.basename(pex_file), flags=re.IGNORECASE)
+    pex_info = PexInfo.from_pex(pex_file)
+    layout = Layout.identify(pex_file)
+    use_platform_suffix = len(configuration.targets) > 1
+    errors = OrderedDict()  # type: OrderedDict[Manifest, str]
+    for manifest in create_manifests(configuration, name, pex_info, layout):
+        args = [science, "--cache-dir", _science_dir(env, "cache")]
+        if env.PEX_VERBOSE:
+            args.append("-{verbosity}".format(verbosity="v" * env.PEX_VERBOSE))
+        dest_dir = os.path.dirname(os.path.abspath(pex_file))
+        args.extend(
+            [
+                "lift",
+                "--file",
+                "pex={pex_file}".format(pex_file=pex_file),
+                "--file",
+                "configure-binding.py={configure_binding}".format(
+                    configure_binding=os.path.join(
+                        os.path.dirname(__file__), "configure-binding.py"
+                    )
+                ),
+                "build",
+                "--dest-dir",
+                dest_dir,
+            ]
+        )
+        if use_platform_suffix:
+            args.append("--use-platform-suffix")
+        args.append(manifest.path)
+        with open(os.devnull, "wb") as devnull:
+            process = subprocess.Popen(args=args, stdout=devnull, stderr=subprocess.PIPE)
```

Comment:
I'm getting permissions error somewhere around here, I had to manually give `science` a `+x`

Reply:
Yup - that's a bug. I already have science on my machine so this code path did not get stressed locally. Fixed.